### PR TITLE
feat: responsive table

### PR DESCRIPTION
**Pull request by vbudovski**
### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
Has been requested/attempted several times.
https://github.com/ant-design/ant-design/issues/11817

### 💡 Background and solution

Large Ant tables do not display well on smaller screens. Allow selective display of columns based on media query check in React.

### 📝 Changelog
Columns are only filtered out in the presence of a responsive attribute.

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |  Table support responsive columns.  |
| 🇨🇳 Chinese | Table 支持响应式展现列。 |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️ 

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

-----
[View rendered CHANGELOG.en-US.md](https://github.com/vbudovski/ant-design/blob/feat/responsive-table/CHANGELOG.en-US.md)
[View rendered CHANGELOG.zh-CN.md](https://github.com/vbudovski/ant-design/blob/feat/responsive-table/CHANGELOG.zh-CN.md)
[View rendered components/empty/index.en-US.md](https://github.com/vbudovski/ant-design/blob/feat/responsive-table/components/empty/index.en-US.md)
[View rendered components/form/index.en-US.md](https://github.com/vbudovski/ant-design/blob/feat/responsive-table/components/form/index.en-US.md)
[View rendered components/form/index.zh-CN.md](https://github.com/vbudovski/ant-design/blob/feat/responsive-table/components/form/index.zh-CN.md)
[View rendered components/select/index.en-US.md](https://github.com/vbudovski/ant-design/blob/feat/responsive-table/components/select/index.en-US.md)

### CODE REVIEW

**Comment by yoyo837** [comment]
Commits should be clean!

**Comment by zombieJ** [comment]
Grid already realize the responsive detect. Should not re-import another lib: https://github.com/ant-design/ant-design/blob/cff24d5dcb03207bf587d454e56bd2d477c0b555/components/grid/row.tsx

**Comment by vbudovski** [comment]
@zombieJ Fixed. It's a lot cleaner now.

File: components/table/demo/responsive.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+---
+order: 31
+title:
+  zh-CN: Responsive
```
Comment (zombieJ) [inline components/table/demo/responsive.md:4]: `Responsive` => `响应式`

File: components/table/demo/responsive.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  en-US: Responsive
+---
+
+## zh-CN
+
+Responsive columns.
```
Comment (zombieJ) [inline components/table/demo/responsive.md:10]: > 响应式配置列的展示。

File: components/table/Table.tsx [hunk 3]
```diff
@@ -375,7 +380,7 @@ function Table<RecordType extends object = any>(props: TableProps<RecordType>) {
     (innerColumns: ColumnsType<RecordType>): ColumnsType<RecordType> => {
       return transformTitleColumns(
         transformSelectionColumns(transformFilterColumns(transformSorterColumns(innerColumns))),
-      );
+      ).filter(c => !c.responsive || c.responsive.filter((r: Breakpoint) => matched.has(r)).length);
```
Comment (zombieJ) [inline components/table/Table.tsx:383]: Move this to the begining.

```tsx
const mergedColumn = React.useMemo(() => { .... }, [columns, children, ...]);
```

And we can now remove `children` parse to columns in `useSorter` & `useSelection`.

Comment (vbudovski) [reply components/table/Table.tsx]: I've moved the memo up top as you suggested. For some reason the responsive display has broken though. I've checked that `mergedColumn` returns the filtered set. Any ideas?

File: components/table/Table.tsx [hunk 4]
```diff
@@ -113,6 +117,16 @@ function Table<RecordType extends object = any>(props: TableProps<RecordType>) {
     showSorterTooltip = true,
   } = props;
 
+  const screens = useBreakpoint();
+  const matched = new Set(Object.keys(screens).filter((m: Breakpoint) => screens[m]));
+
+  const mergedColumns = React.useMemo(() => {
+    return (columns || convertChildrenToColumns(children)).filter(
+      (c: ColumnType<RecordType>) =>
+        !c.responsive || c.responsive.filter((r: Breakpoint) => matched.has(r)).length,
```
Comment (zombieJ) [inline components/table/Table.tsx:126]: Use `some` instead of `filter(...).lenght`

File: components/table/Table.tsx [hunk 5]
```diff
@@ -113,6 +117,16 @@ function Table<RecordType extends object = any>(props: TableProps<RecordType>) {
     showSorterTooltip = true,
   } = props;
 
+  const screens = useBreakpoint();
+  const matched = new Set(Object.keys(screens).filter((m: Breakpoint) => screens[m]));
```
Comment (zombieJ) [inline components/table/Table.tsx:121]: `useMemo` for cache this.

**Comment by zombieJ** [comment]
Thanks for your effort. I'm OK with the code, could you help to update table document about this props?

@afc163  @yoyo837, pls help double confirm for the prop name of `responsive`

**Comment by vbudovski** [comment]
I realised that `useBreakpoint` works a bit differently to the other library. We can probably simplify the interface to take a single instance of `Breakpoint` rather than an array. What do you think @zombieJ ?

**Comment by yoyo837** [comment]
@vbudovski You can send a new PR about that to help improve.

**Review by afc163 — APPROVED** [review]

**Review by yoyo837 — APPROVED** [review]

**Comment by afc163** [comment]
@vbudovski Could you fill API documentation about `responsive`?

https://ant.design/components/table/#Column
https://ant.design/components/table-cn/#Column

**Comment by muzea** [comment]
https://github.com/ant-design/ant-design/blob/c9fada3222c4e9350f4a38bff94dbc5e767f68c5/components/table/Table.tsx#L121-L128


Is it inappropriate for children to be dependent? In my case, it will cause a lot of extra render.

**Comment by afc163** [comment]
@muzea send PR to optimize it.

**Comment by zombieJ** [comment]
> Is it inappropriate for children to be dependent? In my case, it will cause a lot of extra render.

`children` is only used when define Columns by jsx. It's empty when use `columns`.

**Pull request merged**